pkg/render/istio.go

```diff
@@ -0,0 +1,162 @@
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/components"
+	rcomp "github.com/tigera/operator/pkg/render/common/components"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	renderistio "github.com/tigera/operator/pkg/render/istio"
+)
+
+type IstioConfig struct {
+	Installation   *operatorv1.InstallationSpec
+	Istio          *operatorv1.Istio
+	IstioNamespace string
+	Resources      *renderistio.IstioResources
+}
+
+var _ Component = &IstioComponent{}
+
+type IstioComponent struct {
+	cfg                  *IstioConfig
+	IstioPilotImage      string
+	IstioInstallCNIImage string
+	IstioZtunnelImage    string
+	IstioProxyv2Image    string
+}
+
+const (
+	IstioNamespace              = common.CalicoNamespace
+	IstioReleaseName            = "tigera-istio"
+	IstioIstiodDeploymentName   = "istiod"
+	IstioCNIDaemonSetName       = "istio-cni-node"
+	IstioZTunnelDaemonSetName   = "ztunnel"
+	IstioOperatorAnnotationMode = "operator.tigera.io/istioAmbientMode"
+	IstioOperatorAnnotationDSCP = "operator.tigera.io/istioDSCPMark"
+	IstioFinalizer              = "operator.tigera.io/tigera-istio"
+)
+
+func NewIstioComponent(cfg *IstioConfig) *IstioComponent {
+	return &IstioComponent{
+		cfg: cfg,
+	}
+}
+
+func (c *IstioComponent) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := c.cfg.Installation.Registry
+	path := c.cfg.Installation.ImagePath
+	prefix := c.cfg.Installation.ImagePrefix
+	var err error
+	errMsgs := []string{}
+
+	compPilot := components.ComponentTigeraIstioPilot
+	compCNI := components.ComponentTigeraIstioInstallCNI
+	compZTunnel := components.ComponentTigeraIstioZTunnel
+	compProxy := components.ComponentTigeraIstioProxyv2
+	if c.cfg.Installation.Variant == operatorv1.TigeraSecureEnterprise {
+		compPilot = components.ComponentTigeraIstioPilot
+		compCNI = components.ComponentTigeraIstioInstallCNI
+		compZTunnel = components.ComponentTigeraIstioZTunnel
+		compProxy = components.ComponentTigeraIstioProxyv2
+	}
+	c.IstioPilotImage, err = components.GetReference(compPilot, reg, path, prefix, is)
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+	c.IstioInstallCNIImage, err = components.GetReference(compCNI, reg, path, prefix, is)
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+	c.IstioZtunnelImage, err = components.GetReference(compZTunnel, reg, path, prefix, is)
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+	c.IstioProxyv2Image, err = components.GetReference(compProxy, reg, path, prefix, is)
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+
+	if len(errMsgs) > 0 {
+		return fmt.Errorf("%s", strings.Join(errMsgs, ","))
+	}
+
+	return nil
+}
+
+// Objects implements the Component interface.
+func (c *IstioComponent) Objects() ([]client.Object, []client.Object) {
```

Comment:
We need unit tests for this file and this method in particular. This is where the majority of the functionality should be tested.